src/katsdpk8spoc/workflow_controller.py

```diff
@@ -0,0 +1,203 @@
+            template["container"]["args"] = self.command
+        if self.daemon:
+            template["daemon"] = True
+        if self.resources:
+            template["container"]["resources"] = self.resources
+        if self.when:
+            template["when"] = self.when
+        return template
+
+
+class Telstate(Step):
+
+    def __init__(self):
+        """Telstate step
+        """
+        super().__init__()
+        self.name = "telstate",
+        self.template_name = "telstate-template"
+        self.image = "redis"
+        self.daemon = True
+
+
+class Ingest(Step):
+
+    def __init__(self, step_id: int=None):
+        """Ingest step
+
+        :param step_id: the unique ingest ID to be assigned to this process
+        """
+        super().__init__()
+        self.name = f"ingest{step_id}"
+        self.template_name = "ingest-template"
+        self.arguments = ["./run", "-u", "{{tasks.telstate.ip}}"]
+        self.dependencies = "telstate"
+        self.command = "python"
+        self.image = "cchristelis/ingest:0.5"
+
+
+class Calibrator(Step):
+
+    def __init__(self, step_id: int=None):
+        """Calibrator step
+
+        :param step_id: the unique calibrator ID to be assigned to this process
+        """
+        super().__init__()
+        self.name = f"calibrator{step_id}"
+        self.template_name = "calibrator-template"
+        self.arguments = ["./run", "-u", "{{tasks.telstate.ip}}"]
+        self.dependencies = "telstate"
+        self.command = "python"
+        self.image = "cchristelis/calibrator:0.1"
+        self.resources = {
+            "limits": {
+                "cpu": "500m",
+                "memory": "1Gi",
+                "sdp.kat.ac.za/jellybeans": 1  # Fake resource
+            },
+            "requests": {
+                "cpu": "500m",
+                "memory": "1Gi",
+                "sdp.kat.ac.za/jellybeans": 1
+            }
+        }
+
+
+class BatchSetup(Step):
+
+    def __init__(self):
+        """Batch Setup step. This step waits for sufficient inputs from
+        The Ingest and Calibartor to specify the number of batch processes
+        to start.
+        """
+        super().__init__()
+        self.name = "batch-setup",
+        self.template_name = "batch-setup-template"
+        self.image = "cchristelis/batch_setup:0.4"
+        self.dependencies = "telstate"
+        self.command = "python"
+        self.arguments = ["./run", "-u", "{{tasks.telstate.ip}}"]
+
+
+class Batch(Step):
+
+    def __init__(self, number: int):
+        """The Batch step
+
+        :param number:
+        """
+        super().__init__()
+        self.name = f"batch{number}"
+        self.template_name = "batch-template"
+        self.image = "cchristelis/batch:0.1"
+        self.command = "python"
+        self.arguments = ['./run.py']
+        self.dependencies = "batch-setup"
+        self.when = f"{{tasks.batch-setup.outputs.result}} >= {number}"
+
+
+class ProductControllerWorkflow:
```

Comment:
This is good for now.
Passing in config and picking namespace and service account from config will be based on subarray will be needed.

Reply:
This will be followed up in https://github.com/ska-sa/katsdpk8spoc/issues/5